### PR TITLE
Semantic Criteria API

### PR DESCRIPTION
**Pull request by marcel-zec**
Implementation of Semantic Criteria API.

### CODE REVIEW

File: jopa-impl/src/main/java/cz/cvut/kbss/jopa/query/criteria/PathImpl.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+            query.append("." + attributeName);
+        } else {
+            query.append(attributeName);
+        }
+    }
+
+//    @Override
+//    public void setExpressionToQuery(StringBuilder query, CriteriaParameterFiller parameterFiller) {
+//        this.pathSource.setExpressionToQuery(query, parameterFiller);
+//    }
```
Comment (ledsoft) [inline jopa-impl/src/main/java/cz/cvut/kbss/jopa/query/criteria/PathImpl.java:33]: Please, remove this commented-out code

File: jopa-impl/src/main/java/cz/cvut/kbss/jopa/query/criteria/CriteriaParameterFiller.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * Sets value from literal expressions registreted as parameters to query parameters.
+     * @param query - TypedQuery fom setting parameters value
+     */
+    public <T> void setValuesToRegisteredParameters(TypedQueryImpl<T> query) {
+        for(String name: literalParameters.keySet()){
+            ExpressionLiteralImpl<?> parameter = literalParameters.get(name);
+            if(parameter.getLanguageTag() != null) query.setParameter(name, (String) literalParameters.get(name).getValue(), parameter.getLanguageTag());
+            else query.setParameter(name, literalParameters.get(name).getValue());
```
Comment (ledsoft) [inline jopa-impl/src/main/java/cz/cvut/kbss/jopa/query/criteria/CriteriaParameterFiller.java:53]: Please, try to use a more consistent parentheses policy. Sometimes you have if(-else) with parentheses, sometimes without.  Generally, I prefer parentheses to be present.

File: jopa-impl/src/main/java/cz/cvut/kbss/jopa/query/criteria/expressions/ExpressionNotLikeImpl.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package cz.cvut.kbss.jopa.query.criteria.expressions;
+
+import cz.cvut.kbss.jopa.sessions.CriteriaBuilder;
+
+public class ExpressionNotLikeImpl extends AbstractComparisonExpression {
```
Comment (ledsoft) [inline jopa-impl/src/main/java/cz/cvut/kbss/jopa/query/criteria/expressions/ExpressionNotLikeImpl.java:5]: Can't this just be a negated `ExpressionLikeImpl`?

File: jopa-impl/src/test/java/cz/cvut/kbss/jopa/query/criteria/CriteriaBuilderTest.java [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+        final MetamodelImpl metamodel = mock(MetamodelImpl.class);
+        new MetamodelMocks().setMocks(metamodel);
+        final MetamodelProvider mpp = mock(MetamodelProvider.class);
+        when(uowMock.getMetamodel()).thenReturn(metamodel);
+        when(mpp.getMetamodel()).thenReturn(metamodel);
+        when(metamodel.getEntities()).thenReturn(Collections.emptySet());
+        when(mpp.isEntityType(any())).thenAnswer(inv -> metamodel.isEntityType(inv.getArgument(0)));
+
+        f = new CriteriaBuilderImpl(uowMock);
+//        criteriaParameterFiller = new CriteriaParameterFiller();
```
Comment (ledsoft) [inline jopa-impl/src/test/java/cz/cvut/kbss/jopa/query/criteria/CriteriaBuilderTest.java:64]: Please, remove this commented-out code.

**Review by ledsoft — CHANGES_REQUESTED** [review]
See the inline comments.

**Comment by marcel-zec** [comment]
> See the inline comments.

The requested changes have been made in 194842b commit.

**Review by ledsoft — APPROVED** [review]

**Pull request merged**